### PR TITLE
Fix two races

### PR DESCRIPTION
**Pull request by johanneswuerbach**
Fix two races

1. can occur after any connection failure as another `_pulseQueue` is never started and pending items added during the connect won't be processed

2. occurs when a formerly queued client is processed and finds a free pool slot (e.g. because a previous client errored), but during client.connect times out itself.

### CODE REVIEW

**Comment by kibertoad** [comment]
@johanneswuerbach Tests are failing...

File: test/connection-timeout.js [hunk 1]
```diff
@@ -126,4 +127,84 @@ describe('connection timeout', () => {
       })
     })
   })
+
+  it.only('continues processing after a connection failure', (done) => {
```
Comment (kibertoad) [inline test/connection-timeout.js]: was this intended to be committed as .only?

File: test/connection-timeout.js [hunk 2]
```diff
@@ -7,6 +7,7 @@ const describe = require('mocha').describe
 const it = require('mocha').it
 const before = require('mocha').before
 const after = require('mocha').after
+const connectionFailure = new Error('Temporary connection failure');
```
Comment (kibertoad) [inline test/connection-timeout.js]: should this be scoped outside of specific test suite?

Comment (johanneswuerbach) [reply test/connection-timeout.js]: This is re-used by multiple tests, therefor I moved it to the top, but I can also inline it as you prefer.

**Comment by johanneswuerbach** [comment]
Added a fix for the second race.

File: test/connection-timeout.js (hunk 2 above)
Comment (kibertoad) [reply test/connection-timeout.js]: There is no single describe. suite above these tests to which it could be scoped?

**Comment by charmander** [comment]
Can you explain the second one a little more? I’ve copied the simpler first one into a pull request at #111 in case it can to go in earlier.

**Comment by johanneswuerbach** [comment]
@charmander the issue occurs when you use `connectionTimeoutMillis` and in the following situation:

1. connect call A starts connecting to the DB
2. connect call B received between A and successful DB connection is being queue
3. the db connection of call A fails with an error
4. connect call B is pulsed in, finds a free-slot and attempts a database connection
5. connect call B is timed out by `connectionTimeoutMillis` and a failure returned to the client
6. the db connection of B is ready and the callback of connect call B called a 2nd time

My fix now creates a `pendingItem` object, which remembers the callback + timed out state and directly adds the new client to the idle queue in step 6 when the pending item already timed out.

**Comment by charmander** [comment]
Thanks for the explanation!

**Comment by vitaly-t** [comment]
When will we see this merged?

**Comment by johanneswuerbach** [comment]
@brianc any chance you could have a look?

**Comment by brianc** [comment]
I think this looks good! Thanks for the explanation and PR!  If you rebase I'll npm-link it into `pg` and make sure the tests pass over there as well (they should, they don't exercise the pool as much as these do, but wanna be sure) and if everything looks good I'll release a new patch version.

**Comment by johanneswuerbach** [comment]
@brianc done 👍

**Comment by johanneswuerbach** [comment]
Let me also know, if you think https://github.com/brianc/node-pg-pool/pull/110 is worth-while proceeding :-)

**Comment by brianc** [comment]
K - tested this w/ node-postgres. All is in order.  Will release a new patch version shortly.  Thanks again!

**Comment by vitaly-t** [comment]
@brianc Christmas-shortly by any chance? :smile:

**Comment by brianc** [comment]
Ah crap i totally forgot to release this after i tested. Will do! Sorry!

**Comment by brianc** [comment]
yeah...sorry again.  Was out of town.  release `pg-pool@2.0.6`

**Pull request merged**